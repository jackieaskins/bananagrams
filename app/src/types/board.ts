import { Tile } from "./tile";

export type BoardLocation = {
  x: number;
  y: number;
};

export enum Direction {
  ACROSS = "ACROSS",
  DOWN = "DOWN",
}

export enum ValidationStatus {
  NOT_VALIDATED = "NOT_VALIDATED",
  VALID = "VALID",
  INVALID = "INVALID",
}

export type WordInfo = {
  start: BoardLocation;
  validation: ValidationStatus;
};

export type AbstractBoardSquare<T> = {
  tile: T;
  wordInfo: Record<Direction, WordInfo>;
};

export type AbstractBoard<T> = Record<string, AbstractBoardSquare<T>>;

export type BoardSquare = AbstractBoardSquare<Tile>;

export type Board = AbstractBoard<Tile>;
