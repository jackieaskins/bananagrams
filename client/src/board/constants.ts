import { TILE_SIZE } from '../tile/constants';

const SCREEN_MULTIPLIER = 1.5;
const MIN_TILES = 25;

const getNumTiles = (len: number): number =>
  Math.max(Math.ceil((len * SCREEN_MULTIPLIER) / TILE_SIZE), MIN_TILES);
export const TILES_PER_ROW = getNumTiles(window.screen.width);
export const TILES_PER_COL = getNumTiles(window.screen.height);

export const BOARD_WIDTH = TILES_PER_ROW * TILE_SIZE + 2;
export const BOARD_HEIGHT = TILES_PER_COL * TILE_SIZE + 2;

export const BORDER_STYLE = '1px solid rgba(173, 216, 230, 0.5)';
