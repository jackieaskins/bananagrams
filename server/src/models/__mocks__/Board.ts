import Tile from '../Tile';

export default jest.fn().mockImplementation(() => {
  const tiles: Record<string, Tile> = {};

  return {
    toJSON: jest.fn().mockReturnValue('BoardJSON'),
    reset: jest.fn(),
    validateEmptySquare: jest.fn(),
    removeTile: jest
      .fn()
      .mockImplementation((location) => tiles[`${location.x}${location.y}`]),
    addTile: jest.fn().mockImplementation((location, tile) => {
      tiles[`${location.x}${location.y}`] = tile;
    }),
  };
});
