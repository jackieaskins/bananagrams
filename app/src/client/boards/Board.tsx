import { Card, Flex } from "@chakra-ui/react";
import BoardSquare from "./BoardSquare";
import { convertToArray } from "./convert";
import { Board as BoardType } from "@/types/board";

type BoardProps = {
  board: BoardType;
};

export default function Board({ board }: BoardProps): JSX.Element {
  return (
    <Card display="inline-flex" flexDirection="column" variant="outline">
      {convertToArray(board).map((row, y) => (
        <Flex key={y}>
          {row.map((boardSquare, x) => (
            <BoardSquare key={x} x={x} y={y} boardSquare={boardSquare} />
          ))}
        </Flex>
      ))}
    </Card>
  );
}
