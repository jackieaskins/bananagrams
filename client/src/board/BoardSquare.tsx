import { useRecoilCallback } from 'recoil';

import { BoardPosition } from '../boards/types';
import { useCurrentBoardSquare } from '../game/stateHooks';
import { moveTile } from '../socket';
import Tile from '../tile/Tile';
import { TILE_SIZE } from '../tile/constants';
import { selectedTileState } from '../tile/selectedTileState';
import { BORDER_STYLE } from './constants';

const BoardSquare = ({ row, col }: BoardPosition): JSX.Element => {
  const square = useCurrentBoardSquare({ row, col });

  const handleClick = useRecoilCallback(
    ({ set, snapshot }) => async () => {
      const selectedTile = await snapshot.getPromise(selectedTileState);

      if (selectedTile) {
        moveTile({
          tileId: selectedTile.tile.id,
          fromPosition: selectedTile.boardPosition,
          toPosition: { row, col },
        });

        set(selectedTileState, null);
      }
    },
    [row, col]
  );

  const selectTile = useRecoilCallback(
    ({ set, snapshot }) => async ({ tile }) => {
      const selectedTile = await snapshot.getPromise(selectedTileState);

      if (!selectedTile) {
        set(selectedTileState, { tile, boardPosition: { row, col } });
      }
    },
    [col, row]
  );

  return (
    <div
      css={{
        minWidth: `${TILE_SIZE}px`,
        width: `${TILE_SIZE}px`,
        height: `${TILE_SIZE}px`,
        border: BORDER_STYLE,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
      }}
      onClick={handleClick}
    >
      {square ? (
        <Tile
          id={square.tile.id}
          letter={square.tile.letter}
          onClick={selectTile}
          validationStatus={square.validationStatus}
        />
      ) : null}
    </div>
  );
};

export default BoardSquare;
