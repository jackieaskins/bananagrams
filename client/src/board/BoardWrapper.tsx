import { useWindowSize } from '../util/window';
import Board from './Board';
import { BOARD_HEIGHT, BOARD_WIDTH } from './constants';

const BoardView = (): JSX.Element => {
  const { width: windowWidth, height: windowHeight } = useWindowSize();

  return (
    <div
      css={{
        position: 'fixed',
        left:
          BOARD_WIDTH > windowWidth
            ? windowWidth - BOARD_WIDTH - 5
            : (windowWidth - BOARD_WIDTH) / 2,
        right:
          BOARD_WIDTH > windowWidth
            ? windowWidth - BOARD_WIDTH - 5
            : (windowWidth - BOARD_WIDTH) / 2,
        top:
          BOARD_HEIGHT > windowHeight
            ? windowHeight - BOARD_HEIGHT - 5
            : (windowHeight - BOARD_HEIGHT) / 2,
        bottom:
          BOARD_HEIGHT > windowHeight
            ? windowHeight - BOARD_HEIGHT - 5
            : (windowHeight - BOARD_HEIGHT) / 2,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center',
      }}
    >
      <Board />
    </div>
  );
};

export default BoardView;
