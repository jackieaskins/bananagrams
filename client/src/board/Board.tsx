import { useMemo } from 'react';
import Draggable from 'react-draggable';

import BoardSquare from './BoardSquare';
import {
  BOARD_WIDTH,
  BOARD_HEIGHT,
  BORDER_STYLE,
  TILES_PER_COL,
  TILES_PER_ROW,
} from './constants';

const Board = (): JSX.Element => {
  const boardSquares = useMemo(
    () =>
      [...Array(TILES_PER_COL)].map((_, y) => (
        <div key={y} css={{ display: 'flex' }}>
          {[...Array(TILES_PER_ROW)].map((_, x) => (
            <BoardSquare key={x} />
          ))}
        </div>
      )),
    []
  );

  return (
    <Draggable bounds="parent" cancel=".no-drag">
      <div
        css={{
          cursor: 'move',
          display: 'flex',
          flexDirection: 'column',
          width: `${BOARD_WIDTH}px`,
          height: `${BOARD_HEIGHT}px`,
          border: BORDER_STYLE,
        }}
      >
        {boardSquares}
      </div>
    </Draggable>
  );
};

export default Board;
