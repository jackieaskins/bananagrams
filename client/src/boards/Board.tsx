/* eslint-disable @typescript-eslint/ban-ts-comment */
import { Box } from "@mui/material";

import BoardSquare from "./BoardSquare";
import { Board as BoardType } from "../boards/types";
import TransparentPaper from "../paper/TransparentPaper";

type BoardProps = {
  board: BoardType;
};

const Board: React.FC<BoardProps> = ({ board }) => (
  <TransparentPaper
    component={Box}
    // @ts-ignore
    display="inline-flex"
    flexDirection="column"
  >
    {board.map((row, x) => (
      <Box key={x} display="flex">
        {row.map((boardSquare, y) => (
          <BoardSquare key={y} x={x} y={y} boardSquare={boardSquare} />
        ))}
      </Box>
    ))}
  </TransparentPaper>
);

export default Board;
