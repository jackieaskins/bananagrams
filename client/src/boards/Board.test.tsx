import { shallow } from 'enzyme';
import React from 'react';

import { boardSquareFixture } from '../fixtures/board';
import Board from './Board';

describe('<Board />', () => {
  test('renders properly', () => {
    const board = [
      [boardSquareFixture(), null],
      [null, null],
    ];

    expect(shallow(<Board board={board} />)).toMatchSnapshot();
  });
});
