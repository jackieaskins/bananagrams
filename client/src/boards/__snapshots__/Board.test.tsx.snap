// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board /> renders properly 1`] = `
<TransparentPaper
  component={
    {
      "$$typeof": Symbol(react.forward_ref),
      "propTypes": {
        "children": [Function],
        "component": [Function],
        "sx": [Function],
      },
      "render": [Function],
    }
  }
  display="inline-flex"
  flexDirection="column"
>
  <ForwardRef(Box)
    display="flex"
    key="0"
  >
    <BoardSquare
      boardSquare={
        {
          "tile": {
            "id": "A1",
            "letter": "A",
          },
          "wordInfo": {
            "ACROSS": {
              "start": {
                "x": 0,
                "y": 0,
              },
              "validation": "VALID",
            },
            "DOWN": {
              "start": {
                "x": 0,
                "y": 0,
              },
              "validation": "NOT_VALIDATED",
            },
          },
        }
      }
      key="0"
      x={0}
      y={0}
    />
    <BoardSquare
      boardSquare={null}
      key="1"
      x={0}
      y={1}
    />
  </ForwardRef(Box)>
  <ForwardRef(Box)
    display="flex"
    key="1"
  >
    <BoardSquare
      boardSquare={null}
      key="0"
      x={1}
      y={0}
    />
    <BoardSquare
      boardSquare={null}
      key="1"
      x={1}
      y={1}
    />
  </ForwardRef(Box)>
</TransparentPaper>
`;
